File: hanoi.py
#!/usr/bin/python3
#
# $Id: hanoi.py,v 1.1 2016/01/22 22:55:35 urs Exp $

import sys

def usage():
    print("Usage:", sys.argv[0], "n")

def hanoi(n, src, dst, tmp):
    if n == 0:
        return;

    hanoi(n - 1, src, tmp, dst)
    print(src, "->", dst)
    hanoi(n - 1, tmp, dst, src)

if len(sys.argv) != 2:
    usage()
    exit(1)

n = int(sys.argv[1])
hanoi(n, 0, 1, 2)

File: hanoi.c
/*
 * $Id: hanoi.c,v 1.1 2016/01/22 06:57:43 urs Exp $
 */

#include <stdlib.h>
#include <stdio.h>

static void usage(const char *name)
{
	fprintf(stderr, "Usage: %s n src dst\n", name);
}

static void hanoi(int n, int src, int dst);

int main(int argc, char **argv)
{
	int n, src, dst;

	if (argc != 4) {
		usage(argv[0]);
		exit(1);
	}

	n   = atoi(argv[1]);
	src = atoi(argv[2]);
	dst = atoi(argv[3]);

	hanoi(n, src, dst);

	return 0;
}

static void hanoi(int n, int src, int dst)
{
	if (n == 1)
		printf("%d -> %d\n", src, dst);
	else {
		int tmp = 3 - src - dst;
		hanoi(n - 1, src, tmp);
		printf("%d -> %d\n", src, dst);
		hanoi(n - 1, tmp, dst);
	}
}
